lib/mongo_mapper/plugins/dirty.rb

```diff
@@ -46,13 +50,35 @@ def write_key(key, value)
         if !keys.key?(key)
           super
         else
-          attribute_will_change!(key) unless attribute_changed?(key)
+          _attribute_will_change!(key, value)
           super.tap do
-            changed_attributes.delete(key) unless attribute_value_changed?(key)
+            delete_changed_attributes(key) unless attribute_value_changed?(key)
           end
         end
       end
 
+      def _attribute_will_change!(key, value)
+        if ::ActiveModel::VERSION::MAJOR > 4
+          attribute_will_change!(key) if attribute_should_change?(key, value)
+        else
+          attribute_will_change!(key) unless attribute_changed?(key)
+        end
+      end
+
+      def delete_changed_attributes(key)
+        if ::ActiveModel::VERSION::MAJOR > 4
+          return clear_attribute_changes([key])
+        end
+        changed_attributes.delete(key)
+      end
+
+      def attribute_should_change?(key, value)
+        key_val = read_key(key)
+        value != key_val &&
+          (key_val.blank? ? value.present? : true) &&
```

Comment:
This can have a problem when `key_val` is present and `value` is falsey.

For example, this line will return `false` on the following:

Case 1:
- key_val: `nil`
- value: `false`

Case 2:
- key_val: `nil`
- value: `{}`

Reply:
Ah I tried to trace this back to the original error I found the first time I worked on this.
Apparently I mistaken the error was came from the attribute changes condition, it is not. There is no need to change the `attribute_will_change!` invocation here, so I'll just remove these guards entirely.

Thank you for the review @kjcpaas :sparkles: :bowing_man: